core/classes/Adapter.js

```diff
@@ -200,7 +217,79 @@ module.exports = class Adapter {
       }
     })
 
-    return trx ? query.transacting(trx) : query
+    let result = await (trx ? query.transacting(trx) : query)
+    for (let field of otmPopulate) {
+      await this._oneToModelPopulate(result, field)
+    }
+    for (let field of mtmPopulate) {
+      await this._modelToModelPopulate(result, field.linkerModel, field.initialModel, field.modelToLink)
+    }
+    return result
+  }
+
+  /**
+   * Populates the given models with 1:m relation
+   * @param {Object[]} initialModelArr The array of models to be populated
+   * @param {String} populatingModelIdentity The identity of related model
+   * @private
+   */
+  async _oneToModelPopulate(initialModelArr, populatingModelIdentity) {
+    let idsSet = new Set()
+    for (let model of initialModelArr) {
+      idsSet.add(model[populatingModelIdentity])
+    }
+    const idsArr = [...idsSet]
+
+    const populatingModelsArr = await this.knex(populatingModelIdentity.toLowerCase()).whereIn("id", idsArr)
+    for (let model of initialModelArr) {
+      for (let populatingModel of populatingModelsArr) {
```

Comment:
The algorithm can be improved. Before the cycle, create the back index of initial items. It's keys will be the ids of populatingModels, values – lists of the initial models, like this:
```
// items array
[
  { id:1, profile: 1 },
  { id:2, profile: 2 },
  { id:3, profile: 1 },
]
// items hash:
{
  1: [
    { id:1, profile: 1 },
    { id:3, profile: 1 },
  ],
  2: [
    { id:2, profile: 2 }
  ]
}
```
That will let you to avoid inserted cycle and make algorithm a way faster.